trino/auth.py

```diff
@@ -141,102 +143,86 @@ class _OAuth2TokenBearer(AuthBase):
-        GET_TOKEN = 1
+    _BEARER_PREFIX = re.compile(r"bearer", flags=re.IGNORECASE)
 
     def __init__(self, http_session, redirect_auth_url_handler=handle_redirect_auth_url):
         self._redirect_auth_url = redirect_auth_url_handler
-        self.http_session = http_session
         self._thread_local = threading.local()
+        http_session.hooks['response'].append(self.__authenticate)
 
     def __call__(self, r):
-        if not hasattr(self._thread_local, 'init'):
-            self._thread_local.init = True
-            self._thread_local.token = None
-            self._thread_local.token_server = None
-            self._thread_local.attempts = 0
-            self._thread_local.auth_step = self._AuthStep.GET_REDIRECT_SERVER
-
-        if self._thread_local.token:
+        if hasattr(self._thread_local, 'token') and self._thread_local.token:
             r.headers['Authorization'] = "Bearer " + self._thread_local.token
 
         r.register_hook('response', self.__authenticate)
 
         return r
 
-    def __authenticate(self, r, **kwargs):
-        if (self._thread_local.auth_step == self._AuthStep.GET_REDIRECT_SERVER):
-            self.__process_get_redirect_server(r)
-        elif (self._thread_local.auth_step == self._AuthStep.GET_TOKEN):
-            self.__process_get_token(r)
-        return r
-
-    def __process_get_redirect_server(self, r):
-        if not 400 <= r.status_code < 500:
-            return r
+    def __authenticate(self, response, **kwargs):
+        if not 400 <= response.status_code < 500:
+            return response
 
         # we have to handle the authentication, may be token the token expired or it wasn't there at all
-        auth_info = r.headers.get('WWW-Authenticate')
+        auth_info = response.headers.get('WWW-Authenticate')
         if not auth_info:
             raise exceptions.TrinoAuthError("Error: header WWW-Authenticate not available in the response.")
 
-        matches = dict((s.strip().split('=', 1) for s in auth_info.split(",")))
-        if matches is None:
+        if not _OAuth2TokenBearer._BEARER_PREFIX.match(auth_info):
             raise exceptions.TrinoAuthError(f"Error: header info didn't match {auth_info}")
 
-        auth_server = matches.get('Bearer x_redirect_server')
+        matches = parse_dict_header(_OAuth2TokenBearer._BEARER_PREFIX.sub("", auth_info, count=1))
+
+        auth_server = matches.get('x_redirect_server')
         if auth_server is None:
-            raise exceptions.TrinoAuthError("Error: header info didn't have Bearer x_redirect_server")
+            raise exceptions.TrinoAuthError("Error: header info didn't have x_redirect_server")
 
         token_server = matches.get('x_token_server')
         if token_server is None:
             raise exceptions.TrinoAuthError("Error: header info didn't have x_token_server")
 
-        self._thread_local.token_server = token_server.strip('"')
+        self._thread_local.token_server = token_server
```

Comment:
General comment for the future. A single comment is changing logic and doing cleanup refactors. Would be nice to split the two, so review is easier.